Fluid-structure interaction using acoustics

somu15 opened the pull request:
Currently, capabilities to simulate fluid-structure interaction do not exist. Fluid-structure interaction is implemented using the acoustics formulation, which ignores the nonlinear term in the Navier-Stokes equation and performs an Eulerian analysis. The following kernels/bcs are implemented:

* Acoustic inertia kernel to simulate the double time derivative of the fluid pressure
* An interface kernel between fluid and structure to enforce displacement and pressure/stress continuity
* Free surface fluid BC to simulate gravity waves over the fluid

These capabilities are needed to simulate the dynamic response of advanced reactors with molten fuel, among other applications.

Refs #15715
Closes #15712

lindsayad committed:
Create fluid-structure example. Refs #12462

fdkong committed:
Create fluid structure interaction module

Closes #12462

somu15 committed:
FSI codes #15712

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/600546/) on a219140 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6f113e5b2402fe988aaccbe14e912eaad5f5b21a`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/600575/) on 5b9de4b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15995/site)

This comment will be updated on new commits.

dschwen commented on modules/fluid_structure_interaction/doc/moosedocs.py:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
```
```suggestion
#!/usr/bin/env python3
```

dschwen commented on modules/fluid_structure_interaction/include/auxkernels/WaveHeightAuxKernel.h:
```diff
@@ -0,0 +1,39 @@
+protected:
+  virtual Real computeValue();
+
+  // Pressure input
```
Triple slash syntax `///` please, so this gets picked up by doxygen.

dschwen commented on modules/fluid_structure_interaction/include/interfacekernels/FluidStructureInterface.h:
```diff
@@ -0,0 +1,29 @@
+  const MaterialProperty<Real> & _D;
+  const VariableValue & _neighbor_dotdot;
+  const VariableValue & _neighbor_dotdot_du;
+  Real _component;
```
Not a `Real`! Use an `unsigned int` for this!

somu15 replied:
Ahhh! How did I miss this! Thank you.

dschwen commented on modules/fluid_structure_interaction/include/kernels/AcousticInertia.h:
```diff
@@ -0,0 +1,40 @@
+  const VariableValue * _du_dotdot_du;
+
+  /// The TimeIntegrator
+  TimeIntegrator & _time_integrator;
```
make `const` if possible

dschwen replied:
Seems to be possible, as both `uDotFactorTag ` and `uDotDotFactorTag ` are marked `const`. Som, please make sure you address all comments left in a PR.

dschwen commented on modules/fluid_structure_interaction/include/kernels/AcousticInertia.h:
```diff
@@ -0,0 +1,40 @@
+protected:
+  virtual Real computeQpResidual() override;
+
+  virtual Real computeQpJacobian() override;
```
`virtual` is redundant if `override` is specified

dschwen commented on modules/fluid_structure_interaction/src/auxkernels/WaveHeightAuxKernel.C:
```diff
@@ -0,0 +1,38 @@
+  params.addClassDescription("Calculates the wave heights given pressures.");
+  params.addRequiredCoupledVar("pressure", "pressure variable");
+  params.addRequiredParam<Real>("dens", "fluid density");
+  params.addRequiredParam<Real>("grav", "acceleration due to gravity");
```
there really is no good reason to aggressively abbreviate parameter names. Why not just use `density` and `gravity` here?

dschwen commented on modules/fluid_structure_interaction/src/auxkernels/WaveHeightAuxKernel.C:
```diff
@@ -0,0 +1,38 @@
+  : AuxKernel(parameters),
+    _press(coupledValue("pressure")),
+    _dens(getParam<Real>("dens")),
+    _grav(getParam<Real>("grav"))
```
same for the member variables

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,160 @@
+                             "is always the structure and neighbor is always the"
+                             " fluid.");
+  params.addParam<MaterialPropertyName>("D", "D", "Fluid density.");
+  params.addRequiredParam<Real>("component", "The desired component of displacement.");
```
```suggestion
  params.addRequiredRangeCheckedParam<unsigned int>("component", "component < 3", "The desired component of displacement.");
```

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,160 @@
+    _D(getMaterialProperty<Real>("D")),
+    _neighbor_dotdot(_neighbor_var.uDotDotNeighbor()),
+    _neighbor_dotdot_du(_neighbor_var.duDotDotDuNeighbor()),
+    _component(getParam<Real>("component"))
```
```suggestion
    _component(getParam<unsigned int>("component"))
```

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,160 @@
+FluidStructureInterface::computeQpResidual(Moose::DGResidualType type)
+{
+  Real r = 0;
+  if (_component == 0)
```
use a `switch(_component)` here

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,160 @@
+  {
+    switch (type)
+    {
+
```
remove empty lines after `{`

dschwen replied:
should still be removed

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,160 @@
+
+  Real jac = 0;
+
+  if (_component == 0)
```
switch

dschwen replied:
should still be a nested switch. You added that as commented out code and removed it again

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/1D_fluid_only/1D_fluid_only.i:
```diff
@@ -0,0 +1,110 @@
+# Test for `AcousticInertia` and `Diffusion` kernels with only the fluid domain. The
+#  domain is 1D with length 1m and is subjected to an initial condition composed of
+#  a combination of sine waves. Fluid pressure is recorded at the midpoint of the
+# domain. The recorded fluid pressure should match with analytical results. Because
+#  this implementation is equivalent to solving a 1D wave equation, analytical results
+#  exist.
```
```suggestion
# Test for `AcousticInertia` and `Diffusion` kernels with only the fluid domain. The
# domain is 1D with length 1m and is subjected to an initial condition composed of
# a combination of sine waves. Fluid pressure is recorded at the midpoint of the
# domain. The recorded fluid pressure should match with analytical results. Because
# this implementation is equivalent to solving a 1D wave equation, analytical results
# exist.
```

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/1D_fluid_only/1D_fluid_only.i:
```diff
@@ -0,0 +1,110 @@
+# Dimensions = 1
+# Length = 1 meter
+# Fluid speed of sound = 1 m/s
+# Initial condition = sin(3.141*x) + sin(3*3.141*x) + sin(5*3.141*x) + sin(7*3.141*x) + sin(9*3.141*x)
```
```suggestion
# Initial condition = sin(pi*x) + sin(3*pi*x) + sin(5*pi*x) + sin(7*pi*x) + sin(9*pi*x)
```

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/1D_fluid_only/1D_fluid_only.i:
```diff
@@ -0,0 +1,110 @@
+[Functions]
+  [./initial_cond]
+    type = ParsedFunction
+    value = 'sin(3.141*x) + sin(3*3.141*x) + sin(5*3.141*x) + sin(7*3.141*x) + sin(9*3.141*x)'
```
```suggestion
    value = 'sin(pi*x) + sin(3*pi*x) + sin(5*pi*x) + sin(7*pi*x) + sin(9*pi*x)'
```

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/1D_fluid_struc/1D_fluid_struc.i:
```diff
@@ -0,0 +1,232 @@
+# Test for `FluidStructureInterface_3D` interface kernel. The domain is 1D with 20m
+#  length. The fluid domain is on the right and the structural domain is on the left.
+#   Fluid end is subjected to a 250Hz sine wave with a single peak of amplitude unity.
+#   Structural domain is 4 times as dense as the fluid domain with all other material
+#   properties being the same. Fluid pressure is recorded at the midpoint in the fluid
+#   domain (i.e., at 15m). Structural stress is recorded at the midpoint in the structural
+#   domain (i.e., at 5m). The recorded pressure and stress amplitudes should match
+# with theoretical values.
```
```suggestion
# Test for `FluidStructureInterface_3D` interface kernel. The domain is 1D with 20m
# length. The fluid domain is on the right and the structural domain is on the left.
# Fluid end is subjected to a 250Hz sine wave with a single peak of amplitude unity.
# Structural domain is 4 times as dense as the fluid domain with all other material
# properties being the same. Fluid pressure is recorded at the midpoint in the fluid
# domain (i.e., at 15m). Structural stress is recorded at the midpoint in the structural
# domain (i.e., at 5m). The recorded pressure and stress amplitudes should match
# with theoretical values.
```

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/3D_fluid_struc/3D_fluid_struc.i:
```diff
@@ -0,0 +1,391 @@
+# Test for `FluidStructureInterface_3D` interface kernel. The domain is 3D with lengths
+#  10 X 0.1 X 0.1 meters. The fluid domain is on the right and the structural domain
+# is on the left. Fluid end is subjected to a 250Hz sine wave with a single peak.
+#  Structural domain has the same material properties as the fluid. Interface between
+#   structure and fluid is located at 5.0m in the x-direction. Fluid pressure is recorded
+#   at (5, 0.05, 0.05). Structural stress is also recorded at the same location. Fluid
+#   pressure and structural stress should be almost equal and opposite to each other.
+#
```
```suggestion
# Test for `FluidStructureInterface_3D` interface kernel. The domain is 3D with lengths
# 10 X 0.1 X 0.1 meters. The fluid domain is on the right and the structural domain
# is on the left. Fluid end is subjected to a 250Hz sine wave with a single peak.
# Structural domain has the same material properties as the fluid. Interface between
# structure and fluid is located at 5.0m in the x-direction. Fluid pressure is recorded
# at (5, 0.05, 0.05). Structural stress is also recorded at the same location. Fluid
# pressure and structural stress should be almost equal and opposite to each other.
#
```

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/3D_fluid_struc/3D_fluid_struc.i:
```diff
@@ -0,0 +1,391 @@
+    block = 1
+  [../]
+  [./disp_x]
+    block = '0'
```
```suggestion
    block = 0
```

and everywhere else

dschwen commented on modules/fluid_structure_interaction/test/tests/fsi_acoustics/wave_height_bc/wave_height_bc.i:
```diff
@@ -0,0 +1,136 @@
+# Test for `FluidFreeSurfaceBC` BC with only the fluid domain. The domain is 3D with
+#  lengths 1 X 1 X 0.01 meters. It is subjected to a 2D Gaussian initial condition
+#  with the peak at the midpoint (0.5, 0.5, 0.01). Wave heights are recorded at the
+#  midpoint at different times. The recorded wave heights should match with the values
+#   that are provided.
```
```suggestion
# Test for `FluidFreeSurfaceBC` BC with only the fluid domain. The domain is 3D with
# lengths 1 X 1 X 0.01 meters. It is subjected to a 2D Gaussian initial condition
# with the peak at the midpoint (0.5, 0.5, 0.01). Wave heights are recorded at the
# midpoint at different times. The recorded wave heights should match with the values
# that are provided.
```

dschwen commented on modules/fluid_structure_interaction/unit/src/SampleTest.C:
```diff
@@ -0,0 +1,32 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "gtest/gtest.h"
+
+TEST(MySampleTests, descriptiveTestName)
+{
+  // compare equality
+  EXPECT_EQ(2, 1 + 1);
+  EXPECT_DOUBLE_EQ(2 * 3.5, 1.0 * 8 - 1);
+
+  // compare equality and immediately terminate this test if it fails
+  // ASSERT_EQ(2, 1);
+
+  // this won't run if you uncomment the above test because above assert will fail
+  ASSERT_NO_THROW(1 + 1);
+
+  // for a complete list of assertions and for more unit testing documentation see:
+  // https://github.com/google/googletest/blob/master/googletest/docs/Primer.md
+}
+
+TEST(MySampleTests, anotherTest)
+{
+  EXPECT_LE(1, 2);
+  // ...
+}
```
?

dschwen replied:
Why are you adding an empty sample test here?

somu15 replied:
I think @fdkong created this file.

dschwen replied:
I don't understand why you would say this if the git diff clearly shows you are adding this file in this PR... 😕

dschwen replied:
Is this required as a placeholder to not have zero unit tests?

fdkong replied:
This was created by `stork` automatically. 

I think it can be deleted if we do not like it

dschwen commented on modules/modules.mk:
```diff
@@ -63,8 +64,16 @@ ifeq ($(XFEM),yes)
+ifeq ($(FLUID_STRUCTURE_INTERACTION),yes)
+        TENSOR_MECHANICS            := yes
+        NAVIER_STOKES               := yes
+				FLUID_PROPERTIES            := yes
+				RDG                         := yes
+				HEAT_CONDUCTION             := yes
```
? indentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/605134/) on c2400b4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6f113e5b2402fe988aaccbe14e912eaad5f5b21a`

dschwen commented on modules/fluid_structure_interaction/include/bcs/FluidFreeSurfaceBC.h:
```diff
@@ -0,0 +1,34 @@
+
+class FluidFreeSurfaceBC;
+
+template <>
+InputParameters validParams<FluidFreeSurfaceBC>();
```
We should not add the legacy parameter stuff to new code IMO. @aeslaughter ?

dschwen replied:
And this is the only object where you're doing that. So I suggest to just remove this.

dschwen replied:
along with the forward declaration above, which won't be needed anymore

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,259 @@
+  return jac;
+}
+
+// Real
```
Whats going on here?

dschwen commented on modules/fluid_structure_interaction/src/bcs/FluidFreeSurfaceBC.C:
```diff
@@ -0,0 +1,43 @@
+
+registerMooseObject("FluidStructureInteractionApp", FluidFreeSurfaceBC);
+
+defineLegacyParams(FluidFreeSurfaceBC);
```
remove

dschwen commented on modules/fluid_structure_interaction/include/bcs/FluidFreeSurfaceBC.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "IntegratedBC.h"
+
+class FluidFreeSurfaceBC;
```
remove per previous comment

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/606233/) on 4d957cd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6f113e5b2402fe988aaccbe14e912eaad5f5b21a`

fdkong commented:
@somu15 I would like to take a look once all tests are passed.

somu15 commented:
> @somu15 I would like to take a look once all tests are passed.

Sure, @fdkong I will let you know.

somu15 commented:
@fdkong Test modules unit and Check SQA are failing. Any thoughts? Thanks.

somu15 commented:
@fdkong Also, in what folder should I store the figures for the FSI main page? Is it framework/large_media?

fdkong commented:
> @fdkong Test modules unit and Check SQA are failing. Any thoughts? Thanks.

@aeslaughter has better idea on doc failures.

fdkong commented:
> @fdkong Also, in what folder should I store the figures for the FSI main page? Is it framework/large_media?

content/doc/media inset of FSI. Make sure pictures are not ridiculously  large

aeslaughter commented:
All

> @fdkong Also, in what folder should I store the figures for the FSI main page? Is it framework/large_media?

All figures should be placed in the large_media repository, see #14940

aeslaughter commented:
> @fdkong Test modules unit and Check SQA are failing. Any thoughts? Thanks.

I will add the necessary additions for the doc errors

aeslaughter committed:
Add SQA report files

(refs #15995)

somu15 commented:
> > @fdkong Test modules unit and Check SQA are failing. Any thoughts? Thanks.
> 
> I will add the necessary additions for the doc errors

@aeslaughter There is an SQA doc error. Do you have thoughts on this?

aeslaughter commented on modules/fluid_structure_interaction/doc/config.yml:
```diff
@@ -0,0 +1,29 @@
+        home: /modules/fluid_structure_interaction
+    MooseDocs.extensions.appsyntax:
+        executable: ${MOOSE_DIR}/modules/fluid_structure_interaction
+        hide:
+            framework: !include framework/doc/hidden.yml
```
Remove this and your error will go away

somu15 committed:
Update config.yml

aeslaughter commented:
@somu15 I will need to dig into this other error you are getting, it is something that I haven't messed with in a while and not sure what is happening.

somu15 commented:
> @somu15 I will need to dig into this other error you are getting, it is something that I haven't messed with in a while and not sure what is happening.

Thanks @aeslaughter Maybe I need to remove the stuff in folder doc/content/modules/fluid_structure_interaction/sqa ?

aeslaughter committed:
Remove simple diffusion test

(refs #15995)

aeslaughter committed:
Complete SQA setup

(refs #15995)

somu15 committed:
Update AcousticInertia.h

somu15 commented:
@fdkong @dschwen Can you please take a final look at this?

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,152 @@
+      break;
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](1);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](1);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](2);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](2);
+          break;
+      }
+      break;
+    }
+  }
+  return jac;
```
```suggestion
  switch (type)
  {
    case Moose::ElementElement:
      break;
    case Moose::NeighborNeighbor:
      break;
      
    case Moose::ElementNeighbor:
      return _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
            _normals[_qp](_component);

    case Moose::NeighborElement:
      return _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
  }
```

dschwen replied:
You did not implement that suggestion correctly.

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,152 @@
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](0);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](0);
+          break;
+      }
+      break;
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](1);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](1);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](2);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](2);
+          break;
+      }
+      break;
+    }
+  }
+  return r;
```
```suggestion
  switch (type)
  {
    case Moose::Element: // Element is fluid
      return _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](_component);

    case Moose::Neighbor: // Neighbor is structure
      return _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](_component);
  }
```

dschwen reviewed (changes requested):
Some code cleanup

somu15 committed:
code cleanup

dschwen commented on modules/fluid_structure_interaction/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,152 @@
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](_component);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](_component);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::Element: // Element is fluid
+          r = _test[_i][_qp] * _D[_qp] * _neighbor_dotdot[_qp] * _normals[_qp](_component);
+          break;
+
+        case Moose::Neighbor: // Neighbor is structure
+          r = _test_neighbor[_i][_qp] * -_u[_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+  }
+  return r;
+}
```
Please pay closer attention to my review comments.

dschwen reviewed (changes requested):
Fix the unnecessary switch level.

lindsayad commented:
This commit history is not good and will need to be cleaned up
![Screenshot from 2020-12-03 16-55-03](https://user-images.githubusercontent.com/10248304/101107974-5dc5ca80-3588-11eb-90ce-639df6c9c2bb.png)

fdkong commented:
git cherry-pick, and have something like this:

<img width="666" alt="Screen Shot 2020-12-03 at 7 51 34 PM" src="https://user-images.githubusercontent.com/10508374/101115668-f8320800-35a0-11eb-833f-e17e65735d4a.png">

fdkong commented:
@somu15 Could check if you lose anything https://github.com/idaholab/moose/pull/16385?

If we do not lose anything, I will force push to your repo.

fdkong committed:
Rename fluid_structure_interaction to fsi

since fluid_structure_interaction is too long

lindsayad commented:
Fande, love the clean history!

> On Dec 3, 2020, at 6:54 PM, Fande Kong <notifications@github.com> wrote:
> 
> ﻿
> @somu15 Could check if you lose anything #16385?
> 
> If we do not lose anything, I will force push to your repo.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

somu15 commented:
> @somu15 Could check if you lose anything #16385?
> 
> If we do not lose anything, I will force push that to this PR repo (your PR repo).

@fdkong I don't think we lost anything. This looks good!

fdkong commented:
OK, we have a clean history now. I renamed `fluid_structure_interaction` to `fsi`

fdkong commented:
@somu15 Do we have ` tank.png` and `gravity_waves.png` in this PR? I could not find these files even in the original PR (before I cherry pick).

The documentation is failing because ` tank.png` and `gravity_waves.png` do not exist

dschwen commented on modules/fsi/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,152 @@
+    }
+    case 1:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](_component);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+    case 2:
+    {
+      switch (type)
+      {
+        case Moose::ElementElement:
+          break;
+        case Moose::NeighborNeighbor:
+          break;
+        case Moose::ElementNeighbor:
+          jac = _test[_i][_qp] * _D[_qp] * _phi_neighbor[_j][_qp] * _neighbor_dotdot_du[_qp] *
+                _normals[_qp](_component);
+          break;
+        case Moose::NeighborElement:
+          jac = _test_neighbor[_i][_qp] * -_phi[_j][_qp] * _normals[_qp](_component);
+          break;
+      }
+      break;
+    }
+  }
+  return jac;
+}
```
This code still need to be refactored per my previous suggestion

somu15 replied:
Yes, we will incorporate your suggestions.

somu15 replied:
@dschwen I have incorporated your suggestion. Let me know if further changes are required.

dschwen reviewed (changes requested):
implement suggestions

somu15 committed:
Update index.md

somu15 commented:
> @somu15 Do we have ` tank.png` and `gravity_waves.png` in this PR? I could not find these files even in the original PR (before I cherry pick).
> 
> The documentation is failing because ` tank.png` and `gravity_waves.png` do not exist

@fdkong We don't have these two images in the new version of the docs. I have modified the docs. In the future, I would like to have a new PR with animations for FSI from the simulations I am running now.

fdkong committed:
Disable ASSERT_NO_THROW(1 + 1) in unit test

somu15 committed:
Update FluidStructureInterface.C

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/647077/) on 7b85a30 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df`

somu15 committed:
Update FluidStructureInterface.C

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/647103/) on 138f3e4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df`

dschwen commented on modules/fsi/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,138 @@
+Real
+FluidStructureInterface::computeQpJacobian(Moose::DGJacobianType type)
+{
+  switch (_component)
```
remove the switch. Each case is exactly the same now.

dschwen commented on modules/fsi/src/interfacekernels/FluidStructureInterface.C:
```diff
@@ -0,0 +1,138 @@
+Real
+FluidStructureInterface::computeQpResidual(Moose::DGResidualType type)
+{
+  switch (_component)
```
same here. Remove switch, cases are identical

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/653411/) on 039437b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df`

lindsayad reviewed (changes requested):
Ensuring this doesn't get merged until we have export control approval

lindsayad reviewed (dismissed):
Ensuring this doesn't get merged until we have export control approval

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/660336/) on 37632bd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8a36b0980eee59746f57de0e9059c3fa039cc77`

somu15 committed:
FSI interface update

somu15 committed:
Fix conflicts

hugary1995 commented:
Do you have an estimate of when this will have export control approval?

fdkong commented:
> Do you have an estimate of when this will have export control approval?

It was already approved. We are still working on the final cleanup of the code.

hugary1995 commented:
I haven't looked at the code yet. I assume you use ALE fluid domain, right? Do you have any plan to couple the FSI module with the XFEM module to model embedded fluid-structure interface (Lagrangian solid + Eulerian fluid) without distorting the fluid mesh?

@fdkong Actually, the ElementSubdomainModifier you reviewed several days does exactly that.

fdkong committed:
Moved Alex's 2D example to the right place

fdkong committed:
2D toy example works

Closes #12462

fdkong commented:
> I haven't looked at the code yet. I assume you use ALE fluid domain, right? Do you have any plan to couple the FSI module with the XFEM module to model embedded fluid-structure interface (Lagrangian solid + Eulerian fluid) without distorting the fluid mesh?

Did you mean Immersed boundary method? https://en.wikipedia.org/wiki/Immersed_boundary_method

We do not have that right now but it is definitely interesting.  Do you want to add one?

aeslaughter commented on modules/phase_field/src/kernels/MatReaction.C:
```diff
@@ -9,7 +9,7 @@
 #include "MatReaction.h"
 
-registerMooseObject("PhaseFieldApp", MatReaction);
+registerMooseObject("MooseApp", MatReaction);
```
@fdkong This is why the doc location had changed.

fdkong committed:
Moved MatReaction.{C,h,md} to the framework from phase_field module

fdkong committed:
Moved INSConvectedMesh.{C,h,md} to fsi since it is part of ALE

Renamed it to ConvectedMesh.{C,h,md} since it should for both compressible and incompressible flows

fdkong committed:
Moved CoefReaction.{C,h,md} to the framework

and deleted CoefReaction.{C,h} from MooseTestAPP.

We should not duplicate the code

fdkong commented:
It not easy to get things green

@lindsayad Do you have time to take a review?

loganharbour commented on modules/fsi/include/auxkernels/WaveHeightAuxKernel.h:
```diff
@@ -0,0 +1,39 @@
+  Real _density;
+
+  /// Acceleration due to gravity
+  Real _gravity;
```
```suggestion
  const Real _gravity;
```

loganharbour commented on modules/fsi/include/bcs/FluidFreeSurfaceBC.h:
```diff
@@ -0,0 +1,29 @@
+  virtual Real computeQpJacobian() override;
+
+  /// Ratio of u to du/dn. alpha is the inverse of acceleration due to gravity
+  Real _alpha;
```
```suggestion
  const Real _alpha;
```

loganharbour commented on modules/fsi/include/interfacekernels/CoupledPenaltyInterfaceDiffusion.h:
```diff
@@ -0,0 +1,40 @@
+
+#include "InterfaceKernel.h"
+
+// Forward Declarations
+class CoupledPenaltyInterfaceDiffusion;
+
```
```suggestion
```

loganharbour commented on modules/fsi/include/interfacekernels/FluidStructureInterface.h:
```diff
@@ -0,0 +1,29 @@
+  const MaterialProperty<Real> & _D;
+  const VariableValue & _neighbor_dotdot;
+  const VariableValue & _neighbor_dotdot_du;
+  unsigned int _component;
```
```suggestion
  const unsigned int _component;
```

loganharbour commented on modules/fsi/src/bcs/FluidFreeSurfaceBC.C:
```diff
@@ -0,0 +1,41 @@
+{
+  InputParameters params = IntegratedBC::validParams();
+  params.addClassDescription("Applies a mixed Dirichlet-Neumann BC on the fluid surface.");
+  params.addParam<Real>("alpha", 0.1, "Inverse of the acceleration due to gravity.");
```
addRangeCheckedParam?

loganharbour commented on modules/fsi/src/bcs/FluidFreeSurfaceBC.C:
```diff
@@ -0,0 +1,41 @@
+{
+}
+
+Real
+FluidFreeSurfaceBC::computeQpResidual()
+{
+  return _test[_i][_qp] * _alpha * _u_dotdot[_qp] / 1.;
+}
+
+Real
+FluidFreeSurfaceBC::computeQpJacobian()
+{
+  return _test[_i][_qp] * _alpha * _du_dotdot_du[_qp] * _phi[_j][_qp] / 1.;
+}
```
Why the division by 1?

loganharbour commented on modules/fsi/src/interfacekernels/CoupledPenaltyInterfaceDiffusion.C:
```diff
@@ -0,0 +1,121 @@
+CoupledPenaltyInterfaceDiffusion::computeQpOffDiagJacobian(Moose::DGJacobianType type,
+                                                           unsigned int jvar)
+{
+  Real jac = 0;
+  if (jvar == _primary_coupled_id)
+  {
+    switch (type)
+    {
+      case Moose::ElementElement:
+        jac = _test[_i][_qp] * _penalty * _phi[_j][_qp];
+        break;
+
+      case Moose::NeighborElement:
+        jac = _test_neighbor[_i][_qp] * -_penalty * _phi[_j][_qp];
+        break;
+
+      case Moose::ElementNeighbor:
+      case Moose::NeighborNeighbor:
+        break;
+    }
+  }
+  else if (jvar == _secondary_coupled_id)
+  {
+    switch (type)
+    {
+      case Moose::ElementNeighbor:
+        jac = _test[_i][_qp] * _penalty * -_phi_neighbor[_j][_qp];
+        break;
+
+      case Moose::NeighborNeighbor:
+        jac = _test_neighbor[_i][_qp] * -_penalty * -_phi_neighbor[_j][_qp];
+        break;
+
+      case Moose::ElementElement:
+      case Moose::NeighborElement:
+        break;
+    }
+  }
+
+  return jac;
```
```suggestion
  if (jvar == _primary_coupled_id)
  {
    switch (type)
    {
      case Moose::ElementElement:
        return _test[_i][_qp] * _penalty * _phi[_j][_qp];

      case Moose::NeighborElement:
        return _test_neighbor[_i][_qp] * -_penalty * _phi[_j][_qp];

      case Moose::ElementNeighbor:
      case Moose::NeighborNeighbor:
        break;
    }
  }
  else if (jvar == _secondary_coupled_id)
  {
    switch (type)
    {
      case Moose::ElementNeighbor:
        return _test[_i][_qp] * _penalty * -_phi_neighbor[_j][_qp];

      case Moose::NeighborNeighbor:
        return _test_neighbor[_i][_qp] * -_penalty * -_phi_neighbor[_j][_qp];

      case Moose::ElementElement:
      case Moose::NeighborElement:
        break;
    }
  }

  return 0;
```

loganharbour commented on modules/fsi/src/kernels/AcousticInertia.C:
```diff
@@ -0,0 +1,63 @@
+  InputParameters params = TimeKernel::validParams();
+  params.addClassDescription("Calculates the residual for the inertial force "
+                             "which is the double time derivative of pressure.");
+  params.set<bool>("use_displaced_mesh") = true;
```
Should you suppress use_displaced_mesh here and not let the user set it to false?

loganharbour commented on modules/fsi/src/kernels/AcousticInertia.C:
```diff
@@ -0,0 +1,63 @@
+  return params;
+}
+
+AcousticInertia::AcousticInertia(const InputParameters & parameters)
+  : TimeKernel(parameters),
+    _inv_co_sq(getMaterialProperty<Real>("inv_co_sq")),
+    _time_integrator(*_sys.getTimeIntegrator())
+{
+  _u_dot_old = &(_var.uDotOld());
+  _du_dot_du = &(_var.duDotDu());
+  _du_dotdot_du = &(_var.duDotDotDu());
+
+  addFEVariableCoupleableVectorTag(_time_integrator.uDotFactorTag());
+  addFEVariableCoupleableVectorTag(_time_integrator.uDotDotFactorTag());
+
+  _u_dot_factor = &_var.vectorTagValue(_time_integrator.uDotFactorTag());
+  _u_dotdot_factor = &_var.vectorTagValue(_time_integrator.uDotDotFactorTag());
+}
```
```suggestion
AcousticInertia::AcousticInertia(const InputParameters & parameters)
  : TimeKernel(parameters),
    _inv_co_sq(getMaterialProperty<Real>("inv_co_sq")),
    _u_dot_old(_var.uDotOld()),
    _du_dot_du(_var.duDotDu()),
    _du_dotdot_du(_var.duDotDotDu()),
    _u_dot_factor(_var.vectorTagValue(_sys.getTimeIntegrator()->uDotFactorTag())),
    _u_dotdot_factor(_var.vectorTagValue(_sys.getTimeIntegrator()->uDotDotFactorTag()))
{
  addFEVariableCoupleableVectorTag(_sys.getTimeIntegrator()->uDotFactorTag());
  addFEVariableCoupleableVectorTag(_sys.getTimeIntegrator()->uDotDotFactorTag());
}
```

No need for any of these to be pointers.

Also removed the time integrator ref - it's only used in the constructor

loganharbour commented on modules/fsi/src/kernels/AcousticInertia.C:
```diff
@@ -0,0 +1,63 @@
+  if (_dt == 0)
+    return 0;
+  else
+    return _test[_i][_qp] * _inv_co_sq[_qp] *
+           ((*_u_dotdot_factor)[_qp] + (*_u_dot_factor)[_qp] - (*_u_dot_old)[_qp]);
```
```suggestion
    return _test[_i][_qp] * _inv_co_sq[_qp] *
            _u_dotdot_factor[_qp] + _u_dot_factor[_qp] - _u_dot_old[_qp];
```

loganharbour commented on modules/fsi/src/kernels/AcousticInertia.C:
```diff
@@ -0,0 +1,63 @@
+  if (_dt == 0)
+    return 0;
+  else
+    return _test[_i][_qp] * _inv_co_sq[_qp] * (*_du_dotdot_du)[_qp] * _phi[_j][_qp] +
+           _test[_i][_qp] * _inv_co_sq[_qp] * (*_du_dot_du)[_qp] * _phi[_j][_qp];
```
```suggestion
    return _test[_i][_qp] * _inv_co_sq[_qp] * _du_dotdot_du[_qp] * _phi[_j][_qp] +
           _test[_i][_qp] * _inv_co_sq[_qp] * _du_dot_du[_qp] * _phi[_j][_qp];
```

loganharbour commented on modules/fsi/include/kernels/AcousticInertia.h:
```diff
@@ -0,0 +1,40 @@
+
+private:
+  const MaterialProperty<Real> & _inv_co_sq;
+  const VariableValue * _u_dot_factor;
+  const VariableValue * _u_dotdot_factor;
+  const VariableValue * _u_dot_old;
+  const VariableValue * _du_dot_du;
+  const VariableValue * _du_dotdot_du;
```
```suggestion
  const VariableValue & _u_dot_factor;
  const VariableValue & _u_dotdot_factor;
  const VariableValue & _u_dot_old;
  const VariableValue & _du_dot_du;
  const VariableValue & _du_dotdot_du;
```

loganharbour commented on modules/fsi/include/kernels/AcousticInertia.h:
```diff
@@ -0,0 +1,40 @@
+  const VariableValue * _u_dot_old;
+  const VariableValue * _du_dot_du;
+  const VariableValue * _du_dotdot_du;
+
+  /// The TimeIntegrator
+  const TimeIntegrator & _time_integrator;
```
```suggestion
```

loganharbour commented on modules/fsi/include/kernels/AcousticInertia.h:
```diff
@@ -0,0 +1,40 @@
+
+#include "TimeKernel.h"
+#include "Material.h"
+
+// Forward Declarations
+class TimeIntegrator;
```
```suggestion
```

loganharbour commented on modules/fsi/test/src/base/StorkTestApp.C.app:
```diff
@@ -0,0 +1,62 @@
+//* This file is part of the MOOSE framework
```
Remove this file

loganharbour reviewed (dismissed):
I only looked at the c++.

somu15 committed:
Addressing comments 1

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/726711/) on f45d3d4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15995/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15995/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bb0d9cab45bcf0c883f56ec097c400f43a0f599c`

somu15 committed:
Addressing comments 2

fdkong commented:
If there are no more comments, I am going to merge this next week

somu15 committed:
Renaming FluidStructureInterface

moosebuild commented:
Job [Python 3.6 Tests](https://civet.inl.gov/job/735914/) on eebb509 : invalidated by @fdkong

/bin/bash: /tmp/civet_CentOS-8_build_1/tmp91i6n_i1: No such file or directory

somu15 committed:
Re-adding forcing csv

somu15 committed:
Rename interface docs

jiangwen84 commented on modules/fsi/src/interfacekernels/CoupledPenaltyInterfaceDiffusion.C:
```diff
@@ -0,0 +1,111 @@
+registerMooseObject("FsiApp", CoupledPenaltyInterfaceDiffusion);
+
+InputParameters
+CoupledPenaltyInterfaceDiffusion::validParams()
```
@somu15 I just came across this. Is there any reason that you name it with 'xxxDiffusion`? This seems to be a very generic  interface kernel to enforce continuity of any two variables. Other than diffusion problems, it can also enforce `displacement` variables for example. Maybe this can be put in framework. My 2 cents. Nice work!

somu15 replied:
Thanks, @jiangwen84! @fdkong included this kernel. My thought is this kernel can consider any scalar variable and may require some modification to consider vector variables. Also, I am wondering whether this kernel can handle different physics across the interface.

lindsayad replied:
I told @ke7kto just the other day that I think `PenaltyInterfaceDiffusion` from the `test` directory should be moved up into the framework. This object looks like it works similarly. I agree that something like this should go into the framework
